yarn-project/end-to-end/src/e2e_rollup_native_asset_contract.test.ts

```diff
@@ -0,0 +1,218 @@
+    } = await deployL1Contracts(config.rpcUrl, account, logger);
+
+    config.publisherPrivateKey = Buffer.from(privKey!);
+    config.rollupContract = rollupAddress;
+    config.unverifiedDataEmitterContract = unverifiedDataEmitterAddress;
+
+    registryAddress = getAddress(registryAddress_.toString());
+
+    node = await AztecNode.createAndSync(config);
+    aztecRpcServer = await createAztecRpcServer(2, node);
+    accounts = await aztecRpcServer.getAccounts();
+
+    publicClient = createPublicClient({
+      chain: foundry,
+      transport: http(config.rpcUrl),
+    });
+
+    outbox = getContract({
+      address: getAddress(outboxAddress.toString()),
+      abi: OutboxAbi,
+      publicClient,
+    });
+
+    // Deploy L1 portal
+    walletClient = createWalletClient({
+      account,
+      chain: foundry,
+      transport: http(config.rpcUrl),
+    });
+
+    portalAddress = await deployL1Contract(walletClient, publicClient, RollupNativeAssetAbi, RollupNativeAssetBytecode);
+
+    portalContract = getContract({
+      address: getAddress(portalAddress.toString()),
+      abi: RollupNativeAssetAbi,
+      publicClient,
+      walletClient,
+    });
+  }, 60_000);
+
+  afterEach(async () => {
+    await node?.stop();
+    await aztecRpcServer?.stop();
+  });
+
+  const expectBalance = async (owner: AztecAddress, expectedBalance: bigint) => {
+    const ownerPublicKey = await aztecRpcServer.getAccountPublicKey(owner);
+    const [balance] = await contract.methods.getBalance(pointToPublicKey(ownerPublicKey)).view({ from: owner });
+    logger(`Account ${owner} balance: ${balance}`);
+    expect(balance).toBe(expectedBalance);
+  };
+
+  const pointToPublicKey = (point: Point) => {
+    const x = point.buffer.subarray(0, 32);
+    const y = point.buffer.subarray(32, 64);
+    return {
+      x: toBigIntBE(x),
+      y: toBigIntBE(y),
+    };
+  };
+
+  const deployContract = async (initialBalance = 0n, owner = { x: 0n, y: 0n }) => {
+    logger(`Deploying L2 contract...`);
+    const deployer = new ContractDeployer(RollupNativeAssetContractAbi, aztecRpcServer);
+    const tx = deployer.deploy(initialBalance, owner).send({
+      portalContract: portalAddress,
+    });
+    const receipt = await tx.getReceipt();
+    contract = new Contract(receipt.contractAddress!, RollupNativeAssetContractAbi, aztecRpcServer);
+    await contract.attach(portalAddress);
+
+    await tx.isMined(0, 0.1);
+    await tx.getReceipt();
+    logger('L2 contract deployed');
+    return contract;
+  };
+
+  /**
+   * Milestone 2 - L2 to L1
+   */
+  it('Milestone 2.3: Exit funds from L2 to L1', async () => {
+    const initialBalance = 987n;
+    const withdrawAmount = 654n;
+    const [owner] = accounts;
+
+    await deployContract(initialBalance, pointToPublicKey(await aztecRpcServer.getAccountPublicKey(owner)));
+    await expectBalance(owner, initialBalance);
+
+    const { request: initRequest } = await publicClient.simulateContract({
```

Comment:
why do we need to simulate contract? Is this ACVM stuff (which simulates before creating the kernel proof? For whatever reason I assumed this would be abstracted away)

Reply:
This is viem simulating the call and then executing it.